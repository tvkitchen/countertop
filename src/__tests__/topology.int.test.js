import { dataTypes } from '@tvkitchen/base-constants'
import { Countertop } from '..'
import {
	generateMockAppliance,
	normalizeStreams,
} from '../tools/utils/jest'

jest.mock('kafkajs')

describe('Countertop #integration', () => {
	describe('updateTopology', () => {
		it('Should generate correct simple linear topologies', async () => {
			const countertop = new Countertop()
			const sourceAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const captionAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: [dataTypes.TEXT.ATOM],
			}))
			const topology = await countertop.updateTopology()
			const stations = [
				sourceAppliance,
				captionAppliance,
			]
			expect(normalizeStreams(topology.streams, stations)).toMatchSnapshot()
		})

		it('Should generate correct topologies with multiple sources', async () => {
			const countertop = new Countertop()
			const sourceApplianceA = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const sourceApplianceB = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const captionAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: [dataTypes.TEXT.ATOM],
			}))
			const topology = await countertop.updateTopology()
			const stations = [
				sourceApplianceA,
				sourceApplianceB,
				captionAppliance,
			]
			expect(normalizeStreams(topology.streams, stations)).toMatchSnapshot()
		})

		it('Should generate correct topologies with dual inputs', async () => {
			const countertop = new Countertop()
			const sourceAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const captionAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: [dataTypes.TEXT.ATOM],
			}))
			const imageAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: ['IMAGE.JPEG'],
			}))
			const complexInputAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.TEXT.ATOM, 'IMAGE.JPEG'],
				outputTypes: ['IMAGE.GIF'],
			}))
			const topology = await countertop.updateTopology()
			const stations = [
				sourceAppliance,
				captionAppliance,
				imageAppliance,
				complexInputAppliance,
			]
			expect(normalizeStreams(topology.streams, stations)).toMatchSnapshot()
		})

		it('Should generate correct topologies with dual output appliances', async () => {
			const countertop = new Countertop()
			const sourceAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const complexOutputAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: [dataTypes.TEXT.ATOM, 'IMAGE.JPEG'],
			}))
			const topology = await countertop.updateTopology()
			const stations = [
				sourceAppliance,
				complexOutputAppliance,
			]
			expect(normalizeStreams(topology.streams, stations)).toMatchSnapshot()
		})

		it('Should generate correct topologies with dual outputs', async () => {
			const countertop = new Countertop()
			const sourceAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [],
				outputTypes: [dataTypes.STREAM.CONTAINER],
			}))
			const complexOutputAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.STREAM.CONTAINER],
				outputTypes: [dataTypes.TEXT.ATOM, 'IMAGE.JPEG'],
			}))
			const ocrAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: ['IMAGE.JPEG'],
				outputTypes: [dataTypes.TEXT.ATOM],
			}))
			const sentenceAppliance = countertop.addAppliance(generateMockAppliance({
				inputTypes: [dataTypes.TEXT.ATOM],
				outputTypes: [dataTypes.TEXT.SENTENCE],
			}))
			const topology = await countertop.updateTopology()
			const stations = [
				sourceAppliance,
				complexOutputAppliance,
				ocrAppliance,
				sentenceAppliance,
			]
			expect(normalizeStreams(topology.streams, stations)).toMatchSnapshot()
		})
	})
})
