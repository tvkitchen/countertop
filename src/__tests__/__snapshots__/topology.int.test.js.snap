// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countertop #integration updateTopology Should generate correct simple linear topologies 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
]
`;

exports[`Countertop #integration updateTopology Should generate correct topologies with dual inputs 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 2,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 3,
    "source": 0,
    "tributaryMap": Map {
      "TEXT.ATOM" => Object {
        "mouth": 1,
        "source": 0,
        "tributaryMap": Map {
          "STREAM.CONTAINER" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
      "IMAGE.JPEG" => Object {
        "mouth": 2,
        "source": 0,
        "tributaryMap": Map {
          "STREAM.CONTAINER" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
    },
  },
]
`;

exports[`Countertop #integration updateTopology Should generate correct topologies with dual output appliances 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
]
`;

exports[`Countertop #integration updateTopology Should generate correct topologies with dual outputs 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 2,
    "source": 0,
    "tributaryMap": Map {
      "IMAGE.JPEG" => Object {
        "mouth": 1,
        "source": 0,
        "tributaryMap": Map {
          "STREAM.CONTAINER" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
    },
  },
  Object {
    "mouth": 3,
    "source": 0,
    "tributaryMap": Map {
      "TEXT.ATOM" => Object {
        "mouth": 1,
        "source": 0,
        "tributaryMap": Map {
          "STREAM.CONTAINER" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
    },
  },
  Object {
    "mouth": 3,
    "source": 0,
    "tributaryMap": Map {
      "TEXT.ATOM" => Object {
        "mouth": 2,
        "source": 0,
        "tributaryMap": Map {
          "IMAGE.JPEG" => Object {
            "mouth": 1,
            "source": 0,
            "tributaryMap": Map {
              "STREAM.CONTAINER" => Object {
                "mouth": 0,
                "source": 0,
                "tributaryMap": Map {},
              },
            },
          },
        },
      },
    },
  },
  Object {
    "mouth": 3,
    "source": 0,
    "tributaryMap": Map {
      "TEXT.ATOM" => Object {
        "mouth": 2,
        "source": 0,
        "tributaryMap": Map {
          "IMAGE.JPEG" => Object {
            "mouth": 1,
            "source": 0,
            "tributaryMap": Map {
              "STREAM.CONTAINER" => Object {
                "mouth": 0,
                "source": 0,
                "tributaryMap": Map {},
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`Countertop #integration updateTopology Should generate correct topologies with multiple sources 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 1,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 2,
    "source": 0,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 2,
    "source": 1,
    "tributaryMap": Map {
      "STREAM.CONTAINER" => Object {
        "mouth": 1,
        "source": 1,
        "tributaryMap": Map {},
      },
    },
  },
]
`;
